File: backend/src/test/java/com/snackoverflow/toolgether/ReservationControllerTest.java
package com.snackoverflow.toolgether;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import com.snackoverflow.toolgether.domain.reservation.controller.ReservationController;
import com.snackoverflow.toolgether.domain.reservation.dto.ReservationRequest;
import com.snackoverflow.toolgether.domain.reservation.dto.ReservationResponse;
import com.snackoverflow.toolgether.domain.reservation.entity.FailDue;
import com.snackoverflow.toolgether.domain.reservation.service.ReservationService;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.ResultActions;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;

import java.time.LocalDateTime;

import static org.mockito.Mockito.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

public class ReservationControllerTest {

	private MockMvc mockMvc;

	@Mock
	private ReservationService reservationService;

	@InjectMocks
	private ReservationController reservationController;

	private ObjectMapper objectMapper = new ObjectMapper();

	@BeforeEach
	public void setUp() {
		MockitoAnnotations.openMocks(this);
		mockMvc = MockMvcBuilders.standaloneSetup(reservationController).build();
		objectMapper = new ObjectMapper();
		objectMapper.registerModule(new JavaTimeModule());
	}

	@Test
	@DisplayName("예약 요청 테스트")
	public void testRequestReservation() throws Exception {
		ReservationRequest request = new ReservationRequest(1L, 2L, 3L, LocalDateTime.now().plusHours(1), LocalDateTime.now().plusHours(2), 1000.0, 500.0);
		ReservationResponse response = new ReservationResponse(1L, "REQUESTED",  1L, LocalDateTime.now(),  LocalDateTime.now(), 20000.0, "", 3L, 2L);

		when(reservationService.requestReservation(any(ReservationRequest.class))).thenReturn(response);

		ResultActions resultActions = mockMvc.perform(post("/api/v1/reservations/request")
				.contentType(MediaType.APPLICATION_JSON)
				.content(objectMapper.writeValueAsString(request)))
			.andDo(print());

		resultActions
			.andExpect(status().isOk());

		verify(reservationService, times(1)).requestReservation(any(ReservationRequest.class));
	}

	@Test
	@DisplayName("예약 승인 테스트")
	public void testApproveReservation() throws Exception {
		ResultActions resultActions = mockMvc.perform(post("/api/v1/reservations/1/approve"))
			.andDo(print());

		resultActions
			.andExpect(status().isOk());

		verify(reservationService, times(1)).approveReservation(1L);
	}

	@Test
	@DisplayName("예약 거절 테스트")
	public void testRejectReservation() throws Exception {
		ResultActions resultActions = mockMvc.perform(patch("/api/v1/reservations/1/reject")
				.param("reason", "Reason for rejection"))
			.andDo(print());

		resultActions
			.andExpect(status().isOk());

		verify(reservationService, times(1)).rejectReservation(1L, "Reason for rejection");
	}

	@Test
	@DisplayName("대여 시작 변경 테스트")
	public void testStartRental() throws Exception {
		ResultActions resultActions = mockMvc.perform(patch("/api/v1/reservations/1/start"))
			.andDo(print());

		resultActions
			.andExpect(status().isOk());

		verify(reservationService, times(1)).startRental(1L);
	}

	@Test
	@DisplayName("대여 종료 변경 테스트")
	public void testCompleteRental() throws Exception {
		ResultActions resultActions = mockMvc.perform(patch("/api/v1/reservations/1/complete"))
			.andDo(print());

		resultActions
			.andExpect(status().isOk());

		verify(reservationService, times(1)).completeRental(1L);
	}

	@Test
	@DisplayName("소유자 이슈로 인한 환불 테스트")
	public void testOwnerIssue() throws Exception {
		ResultActions resultActions = mockMvc.perform(patch("/api/v1/reservations/1/ownerIssue")
				.param("reason", "Owner issue reason"))
			.andDo(print());

		resultActions
			.andExpect(status().isOk());

		verify(reservationService, times(1)).failDueTo(1L, "Owner issue reason", FailDue.OWNER_ISSUE);
	}

	@Test
	@DisplayName("대여자 이슈로 인한 환불 테스트")
	public void testRenterIssue() throws Exception {
		ResultActions resultActions = mockMvc.perform(patch("/api/v1/reservations/1/renterIssue")
				.param("reason", "Renter issue reason"))
			.andDo(print());

		resultActions
			.andExpect(status().isOk());

		verify(reservationService, times(1)).failDueTo(1L, "Renter issue reason", FailDue.RENTER_ISSUE);
	}
}

File: backend/src/test/kotlin/com/snackoverflow/toolgether/reservation/ReservationControllerTest.kt
package com.snackoverflow.toolgether.reservation

import com.snackoverflow.toolgether.domain.reservation.controller.ReservationController
import jakarta.transaction.Transactional
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.mockito.MockitoAnnotations
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.ResultActions
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*
import org.springframework.test.web.servlet.result.MockMvcResultHandlers
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*
import org.springframework.test.web.servlet.setup.MockMvcBuilders
import org.springframework.web.context.WebApplicationContext
import java.nio.charset.StandardCharsets
import java.time.LocalDateTime

@SpringBootTest
@ActiveProfiles("")
@AutoConfigureMockMvc
@Transactional
class ReservationControllerTest {

    lateinit var mockMvc: MockMvc

    @Autowired
    protected val context: WebApplicationContext? = null

    private val user1 = 1L
    private val user2 = 2L
    private val post1 = 1L

    @BeforeEach
    fun setUp() {
        MockitoAnnotations.openMocks(this)
        mockMvc = MockMvcBuilders.webAppContextSetup(context).build()
    }

    private fun requestReservation(): ResultActions {
        val jsonContent = """
        {
            "postId": ${post1},
            "renterId": ${user2},
            "ownerId": ${user1},
            "startTime": "${LocalDateTime.now().plusDays(1)}",
            "endTime": "${LocalDateTime.now().plusDays(3)}",
            "deposit": 10000.0,
            "rentalFee": 500.0
        }
        """.trimIndent()

        println("Sending JSON: $jsonContent")

        return mockMvc
            .perform(
                post("/api/v1/reservations/request")
                    .content(jsonContent)
                    .contentType(MediaType(MediaType.APPLICATION_JSON, StandardCharsets.UTF_8))
            )
    }

    @Test
    @Transactional
    @DisplayName("예약 요청 테스트")
    fun request(){
        val resultActions = requestReservation().andDo(MockMvcResultHandlers.print())

        resultActions
            .andExpect(status().isOk)
            .andExpect(handler().handlerType(ReservationController::class.java))
            .andExpect(handler().methodName("createReservation"))
            .andExpect(jsonPath("$.code").value("201-1"))
            .andExpect(jsonPath("$.msg").value("예약 요청 성공"))
            .andExpect(jsonPath("$.data.status").value("REQUESTED"))
    }

    @Test
    @DisplayName("예약 승인 테스트")
    fun approve(){
        val resultAction = mockMvc.perform(
            patch("/api/v1/reservations/1/approve")
        ).andDo(MockMvcResultHandlers.print())

        resultAction.andExpect(status().isOk)
            .andExpect(handler().handlerType(ReservationController::class.java))
            .andExpect(handler().methodName("approveReservation"))
            .andExpect(jsonPath("$.code").value("201-1"))
            .andExpect(jsonPath("$.msg").value("1번 예약 승인 성공"))
    }

    @Test
    @Transactional
    @DisplayName("예약 거절 테스트")
    fun reject(){
        val resultAction = mockMvc.perform(
            patch("/api/v1/reservations/1/reject")
        ).andDo(MockMvcResultHandlers.print())

        resultAction.andExpect(status().isOk)
            .andExpect(handler().handlerType(ReservationController::class.java))
            .andExpect(handler().methodName("rejectReservation"))
            .andExpect(jsonPath("$.code").value("200-1"))
            .andExpect(jsonPath("$.msg").value("1번 예약 거절 성공"))
    }

    @Test
    @DisplayName("예약 취소 테스트")
    fun cancel(){
        val resultAction = mockMvc.perform(
            patch("/api/v1/reservations/1/cancel")
        ).andDo(MockMvcResultHandlers.print())

        resultAction.andExpect(status().isOk)
            .andExpect(handler().handlerType(ReservationController::class.java))
            .andExpect(handler().methodName("cancelReservation"))
            .andExpect(jsonPath("$.code").value("200-1"))
            .andExpect(jsonPath("$.msg").value("1번 예약 취소 성공"))
    }

    @Test
    @DisplayName("대여 시작 테스트")
    fun start(){
        val resultActions = requestReservation().andDo(MockMvcResultHandlers.print())

        resultActions
            .andExpect(status().isOk)
            .andExpect(handler().handlerType(ReservationController::class.java))
            .andExpect(handler().methodName("createReservation"))
            .andExpect(jsonPath("$.code").value("201-1"))
            .andExpect(jsonPath("$.msg").value("예약 요청 성공"))
            .andExpect(jsonPath("$.data.status").value("REQUESTED"))

        val resultAction = mockMvc.perform(
            patch("/api/v1/reservations/2/start")
        ).andDo(MockMvcResultHandlers.print())

        resultAction.andExpect(status().isOk)
            .andExpect(handler().handlerType(ReservationController::class.java))
            .andExpect(handler().methodName("startRental"))
            .andExpect(jsonPath("$.code").value("200-1"))
            .andExpect(jsonPath("$.msg").value("2번 예약 대여 시작 성공"))
    }

    @Test
    @DisplayName("소유자 이슈 테스트")
    fun issue1() {
        val resultAction = mockMvc.perform(
            patch("/api/v1/reservations/2/ownerIssue")
                .param("reason", "DAMAGE_REPORTED")
        ).andDo(MockMvcResultHandlers.print())

        resultAction
            .andExpect(status().isOk)
            .andExpect(handler().handlerType(ReservationController::class.java))
            .andExpect(handler().methodName("ownerIssue"))
            .andExpect(jsonPath("$.code").value("200-1"))
            .andExpect(jsonPath("$.msg").value("2번 예약 소유자에 의한 이슈로 환급 성공"))
    }

    @Test
    @DisplayName("대여자 이슈 테스트")
    fun issue2() {
        val resultAction = mockMvc.perform(
            patch("/api/v1/reservations/2/renterIssue")
                .param("reason", "ITEM_LOSS")
        ).andDo(MockMvcResultHandlers.print())

        resultAction
            .andExpect(status().isOk)
            .andExpect(handler().handlerType(ReservationController::class.java))
            .andExpect(handler().methodName("renterIssue")) // 메서드 이름 수정
            .andExpect(jsonPath("$.code").value("200-1"))
            .andExpect(jsonPath("$.msg").value("2번 예약 대여자에 의한 이슈로 환급 성공"))
    }

    @Test
    @DisplayName("예약 번호 조회 테스트")
    fun getReservation() {

        val resultAction = mockMvc.perform(get("/api/v1/reservations/1"))

        val mvcResult = resultAction.andReturn()
        val responseBody = mvcResult.response.contentAsString
        println("Response Body: $responseBody") // 응답 본문 출력

        resultAction
            .andExpect(status().isOk)
            .andExpect(handler().handlerType(ReservationController::class.java))
            .andExpect(handler().methodName("getReservationById")) // 메서드 이름 수정
            .andExpect(jsonPath("$.code").value("200-1"))
            .andExpect(jsonPath("$.msg").value("1번 예약 조회 성공"))
    }

    @Test
    @DisplayName("게시글 번호 조회 테스트")
    fun getReservations() {
        val resultAction = mockMvc.perform(
            get("/api/v1/reservations/reservatedDates/1")
        )

        val mvcResult = resultAction.andReturn()
        val responseBody = mvcResult.response.contentAsString
        println("Response Body: $responseBody") // 응답 본문 출력

        resultAction
            .andExpect(status().isOk)
            .andExpect(handler().handlerType(ReservationController::class.java))
            .andExpect(handler().methodName("getReservedDates")) // 메서드 이름 수정
            .andExpect(jsonPath("$.code").value("200-1"))
            .andExpect(jsonPath("$.msg").value("1번 게시글의 예약 일정 조회 성공"))
    }
}